ddtrace/tracer/handler.go

```diff
@@ -0,0 +1,167 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+package tracer
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+type traceHandler interface {
+	// pushPayload adds a trace to the current payload being constructed by the handler.
+	pushPayload([]*span)
+
+	// flush should cause the handler to send its current payload. The flush can happen asynchronously, but
+	// the handler must be ready to accept new traces with pushPayload when flush() returns.
+	flush()
+
+	// wait will pause until the traceHandler is finished any asynchronous tasks it is performing.
+	wait()
+
+	// sample gives a handler the opportunity to perform sampling on a span. This is called when a span
+	// is created.
+	sample(*span)
+}
+
+type agentTraceHandler struct {
+	config *config
+
+	payload *payload
+
+	// climit limits the number of concurrent outgoing connections
+	climit chan struct{}
+
+	// wg waits for all goroutines to exit when stopping.
+	wg sync.WaitGroup
+
+	// prioritySampling refers to the tracer's priority sampler.
+	prioritySampling *prioritySampler
+}
+
+var _ traceHandler = &agentTraceHandler{}
+
+func (h *agentTraceHandler) pushPayload(trace []*span) {
+	if err := h.payload.push(trace); err != nil {
+		h.config.statsd.Incr("datadog.tracer.traces_dropped", []string{"reason:encoding_error"}, 1)
+		log.Error("error encoding msgpack: %v", err)
+	}
+	if h.payload.size() > payloadSizeLimit {
+		h.config.statsd.Incr("datadog.tracer.flush_triggered", []string{"reason:size"}, 1)
+		h.flush()
+	}
+}
+
+// flush will push any currently buffered traces to the server.
+func (h *agentTraceHandler) flush() {
+	if h.payload.itemCount() == 0 {
+		return
+	}
+	h.wg.Add(1)
+	h.climit <- struct{}{}
+	go func(p *payload) {
+		defer func(start time.Time) {
+			<-h.climit
+			h.wg.Done()
+			h.config.statsd.Timing("datadog.tracer.flush_duration", time.Since(start), nil, 1)
+		}(time.Now())
+		size, count := p.size(), p.itemCount()
+		log.Debug("Sending payload: size: %d traces: %d\n", size, count)
+		rc, err := h.config.transport.send(p)
+		if err != nil {
+			h.config.statsd.Count("datadog.tracer.traces_dropped", int64(count), []string{"reason:send_failed"}, 1)
+			log.Error("lost %d traces: %v", count, err)
+		} else {
+			h.config.statsd.Count("datadog.tracer.flush_bytes", int64(size), nil, 1)
+			h.config.statsd.Count("datadog.tracer.flush_traces", int64(count), nil, 1)
+			if err := h.prioritySampling.readRatesJSON(rc); err != nil {
+				h.config.statsd.Incr("datadog.tracer.decode_error", nil, 1)
+			}
+		}
+	}(h.payload)
+	h.payload = newPayload()
+}
+
+func (h *agentTraceHandler) sample(s *span) {
+	h.prioritySampling.apply(s)
+}
+
+func (h *agentTraceHandler) wait() {
+	h.wg.Wait()
+}
+
+type hexInt uint64
+
+func (i hexInt) MarshalJSON() ([]byte, error) {
+	return []byte(fmt.Sprintf(`"%X"`, i)), nil
```

Comment:
This is going to be slow. Does it matter?

Reply:
I'll try `strconv.FormatUint` instead. Should be faster.